Wrong shifting behaviour in MetaSMT

Hi @hoangmle, thanks for your contribution #409 .
Unluckily there is a bug, could you please have a look at the following test case:
https://github.com/klee/klee/blob/master/test/regression/2016-04-14-sdiv-2.c_

Rename it by removing the "_" and remove the XFail line.

I get the following error for LLVM 3.4.:
```
Command 2 Stderr:

KLEE: output directory is "/home/travis/build/klee/build/klee/test/regression/Output/2016-04-14-sdiv-2.c.tmp.klee-out"

Using MetaSMT solver backend

Starting MetaSMTSolver(Boolector) ...

KLEE: WARNING: undefined reference to function: printf

terminate called after throwing an instance of 'metaSMT::solver::Boolector::BoolectorAssertion'

  what():  internal error in boolector

0  klee            0x0000000000eca422 llvm::sys::PrintStackTrace(_IO_FILE*) + 34

1  klee            0x0000000000eca7b9

2  libpthread.so.0 0x00007fb6f33fccb0

3  libc.so.6       0x00007fb6f0a2f035 gsignal + 53

4  libc.so.6       0x00007fb6f0a3279b abort + 379

5  libstdc++.so.6  0x00007fb6f133606d __gnu_cxx::__verbose_terminate_handler() + 349

6  libstdc++.so.6  0x00007fb6f1334026

7  libstdc++.so.6  0x00007fb6f1334071

8  libstdc++.so.6  0x00007fb6f1334288

9  klee            0x00000000005e829e metaSMT::solver::Boolector::_boolector_error(int) + 110

10 libboolector.so 0x00007fb6f2d635c4

11 libboolector.so 0x00007fb6f2d660dc boolector_eq + 364

12 klee            0x00000000005e7850

13 klee            0x00000000005f4fc8 klee::MetaSMTBuilder<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::bvVarRightShift(BtorNode*, BtorNode*, unsigned int) + 456

14 klee            0x00000000005f52d2 klee::MetaSMTBuilder<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::constructSDivByConstant(BtorNode*, unsigned int, unsigned long) + 370

15 klee            0x000000000060ce18 klee::MetaSMTBuilder<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::constructActual(klee::ref<klee::Expr>, int*) + 7544

16 klee            0x000000000060dfa8 klee::MetaSMTBuilder<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::construct(klee::ref<klee::Expr>, int*) + 1000

17 klee            0x000000000060b26f klee::MetaSMTBuilder<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::constructActual(klee::ref<klee::Expr>, int*) + 463

18 klee            0x000000000060dfa8 klee::MetaSMTBuilder<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::construct(klee::ref<klee::Expr>, int*) + 1000

19 klee            0x000000000060e1bc klee::MetaSMTBuilder<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::construct(klee::ref<klee::Expr>) + 44

20 klee            0x0000000000611ecf klee::MetaSMTSolverImpl<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::runAndGetCexForked(klee::Query const&, std::vector<klee::Array const*, std::allocator<klee::Array const*> > const&, std::vector<std::vector<unsigned char, std::allocator<unsigned char> >, std::allocator<std::vector<unsigned char, std::allocator<unsigned char> > > >&, bool&, double) + 399

21 klee            0x00000000006134f3 klee::MetaSMTSolverImpl<metaSMT::DirectSolver_Context<metaSMT::solver::Boolector> >::computeInitialValues(klee::Query const&, std::vector<klee::Array const*, std::allocator<klee::Array const*> > const&, std::vector<std::vector<unsigned char, std::allocator<unsigned char> >, std::allocator<std::vector<unsigned char, std::allocator<unsigned char> > > >&, bool&) + 387

22 klee            0x00000000005d2a4c CexCachingSolver::getAssignment(klee::Query const&, klee::Assignment*&) + 268

23 klee            0x00000000005d3328 CexCachingSolver::computeValidity(klee::Query const&, klee::Solver::Validity&) + 952

24 klee            0x00000000005d0f6d CachingSolver::computeValidity(klee::Query const&, klee::Solver::Validity&) + 109

25 klee            0x00000000005e2574 IndependentSolver::computeValidity(klee::Query const&, klee::Solver::Validity&) + 292

26 klee            0x00000000005a7f9f klee::TimingSolver::evaluate(klee::ExecutionState const&, klee::ref<klee::Expr>, klee::Solver::Validity&) + 223

27 klee            0x000000000057a459 klee::Executor::fork(klee::ExecutionState&, klee::ref<klee::Expr>, bool) + 233

28 klee            0x0000000000581573 klee::Executor::executeInstruction(klee::ExecutionState&, klee::KInstruction*) + 8563

29 klee            0x000000000058679e klee::Executor::run(klee::ExecutionState&) + 1374

30 klee            0x0000000000586ff1 klee::Executor::runFunctionAsMain(llvm::Function*, int, char**, char**) + 1697

31 klee            0x0000000000554a2d main + 12525

32 libc.so.6       0x00007fb6f0a1a7ed __libc_start_main + 237

33 klee            0x0000000000568ce5

error: metaSMT did not return successfully (status = 6) 

EXITING ON ERROR:

Query timed out (fork).
```

Thanks a lot for your help!

--- a/lib/Solver/MetaSMTBuilder.h
+++ b/lib/Solver/MetaSMTBuilder.h
@@ -540,6 +540,9 @@ MetaSMTBuilder<SolverContext>::bvVarLeftShift(
     typename SolverContext::result_type expr,
     typename SolverContext::result_type shift, unsigned width) {
 
+  assert(_solver.get_bv_width(expr) == width);
+  assert(_solver.get_bv_width(shift) == width);
+
   typename SolverContext::result_type res = bvZero(width);
 
   // construct a big if-then-elif-elif-... with one case per possible shift
@@ -565,6 +568,9 @@ typename SolverContext::result_type
 MetaSMTBuilder<SolverContext>::bvVarRightShift(
     typename SolverContext::result_type expr,
     typename SolverContext::result_type shift, unsigned width) {
+
+  assert(_solver.get_bv_width(expr) == width);
+  assert(_solver.get_bv_width(shift) == width);
 
   typename SolverContext::result_type res = bvZero(width);
 
@@ -592,6 +598,9 @@ typename SolverContext::result_type
 MetaSMTBuilder<SolverContext>::bvVarArithRightShift(
     typename SolverContext::result_type expr,
     typename SolverContext::result_type shift, unsigned width) {
+
+  assert(_solver.get_bv_width(expr) == width);
+  assert(_solver.get_bv_width(shift) == width);
 
   // get the sign bit to fill with
   typename SolverContext::result_type signedBool =

--- a/lib/Solver/MetaSMTBuilder.h
+++ b/lib/Solver/MetaSMTBuilder.h
@@ -450,7 +450,7 @@ MetaSMTBuilder<SolverContext>::constructSDivByConstant(
   int32_t mprime, dsign, shpost;
   ComputeSDivConstants32(d, mprime, dsign, shpost);
   typename SolverContext::result_type expr_dsign = bvConst32(32, dsign);
-  typename SolverContext::result_type expr_shpost = bvConst32(32, shpost);
+  typename SolverContext::result_type expr_shpost = bvConst32(64, shpost);
 
   // q0 = n + MULSH( mprime, n ) = n + (( (int64_t)mprime * (int64_t)n ) >> 32)
   int64_t mprime_64 = (int64_t)mprime;

